File: incre-tests/synduce/indexed_list/position_polynomial.f
Config ExtraGrammar = "AutoLifter";
Config NonLinear = true;

Inductive List = nil Unit | cons {Int, List};
Inductive IndexedList = inil Unit | icons {Int, Int, IndexedList};

length = fix (
  \f: List -> Int. \xs: List.
  match xs with
    nil _ -> 0
  | cons {h, t} -> + 1 (f t)
  end
);

repr = fix (
  \f: List -> Compress IndexedList. \xs: List.
  match xs with
    nil _ -> inil unit
  | cons {h, t} -> icons {h, length t, f t}
  end
);

spec = fix (
  \f: IndexedList -> Int. \xs: IndexedList.
  match xs with
    inil _ -> 0
  | icons {h, i, t} -> + (* h i) (f t)
  end
);

main = \xs: List. spec (repr xs);

File: ml-benchmark/synduce/indexed_list/position_polynomial.ml
config EnableDeepCoder = true
config NonLinear = true

type nat =
  | Z
  | S of nat

type list =
  | Nil
  | Cons of int * list

type indexed_list =
  | CNil
  | CCons of int * int * indexed_list

let rec length = function
  | Nil -> 0
  | Cons (_, tl) -> 1 + length tl

val repr: list -> indexed_list compress
let rec repr = function
  | Nil -> CNil
  | Cons (value, tl) -> CCons (value, length tl, repr tl)

let rec polynome = function
  | CNil -> 0
  | CCons (hd, i, tl) -> (hd * i) + polynome tl

let prog xs = polynome (repr xs)